### PR TITLE
Initial GraphQL interface implementation

### PR DESCRIPTION
**Pull request by SLoeuillet**
That's an initial PR mostly for code review, not ready for production use

Got basic GraphQL working when querying a single block

### CODE REVIEW

**Comment by SLoeuillet** [comment]
GraphQL actually enabled if using --graphql but could be changed to use --http.api=graphql

To test it, go to 
http://localhost:8545/graphql for request, /graphql/ui for user-friendly UI

A sample block query is bundled in the graphql/ subdirectory

**Comment by AskAlexSharov** [comment]
looks good. need port geth's tests also

**Comment by SLoeuillet** [comment]
@AskAlexSharov : I'll try to add unit tests this week-end

Any preference for protocol activation ? (--graphql to mimic Geth) or via the "erigon standard" --http.api=graphql ?

**Comment by AskAlexSharov** [comment]
> @AskAlexSharov : I'll try to add unit tests this week-end
> 
> Any preference for protocol activation ? (--graphql to mimic Geth) or via the "erigon standard" --http.api=graphql ?

Geth-compatible

File: cmd/rpcdaemon/commands/graphql_api.go [hunk 1]
```diff
@@ -0,0 +1,171 @@
...
+	if b == nil {
+		return nil, nil
+	}
+
+	chainConfig, err := api.chainConfig(tx)
+	if err != nil {
+		return nil, err
+	}
+
+	receipts, err := api.getReceipts(ctx, tx, chainConfig, b, b.Body().SendersFromTxs())
```
Comment (AskAlexSharov) [inline cmd/rpcdaemon/commands/graphql_api.go:59]: Use existing variable “senders”

File: cmd/rpcdaemon/commands/graphql_api.go [hunk 2]
```diff
@@ -0,0 +1,171 @@
...
+
+	response := map[string]interface{}{}
+	response["block"] = getBlockRes
+	response["issuance"] = getIssuanceRes
+	response["totalFees"] = hexutil.Uint64(feesRes)
+	response["receipts"] = result
+	return response, nil
+}
+
+func (api *GraphQLAPIImpl) getBlockWithSenders(ctx context.Context, number rpc.BlockNumber, tx kv.Tx) (*types.Block, []common.Address, error) {
```
Comment (AskAlexSharov) [inline cmd/rpcdaemon/commands/graphql_api.go:90]: BaseApi has similar method already? (It also using lru cache as i remember)

File: cmd/rpcdaemon/commands/graphql_api.go [hunk 3]
```diff
@@ -0,0 +1,171 @@
...
+	return response, err
+}
+
+func (api *GraphQLAPIImpl) delegateIssuance(tx kv.Tx, block *types.Block, chainConfig *chain.Config) (internalIssuance, error) {
+	if chainConfig.Ethash == nil {
+		// Clique for example has no issuance
+		return internalIssuance{}, nil
+	}
+
+	minerReward, uncleRewards := ethash.AccumulateRewards(chainConfig, block.Header(), block.Uncles())
```
Comment (AskAlexSharov) [inline cmd/rpcdaemon/commands/graphql_api.go:135]: Do HeaderNoCopy

**Review by AskAlexSharov — CHANGES_REQUESTED** [review]

File: cmd/rpcdaemon/commands/graphql_api.go (hunk 1 above)
Comment (SLoeuillet) [reply cmd/rpcdaemon/commands/graphql_api.go]: done

File: cmd/rpcdaemon/commands/graphql_api.go (hunk 2 above)
Comment (SLoeuillet) [reply cmd/rpcdaemon/commands/graphql_api.go:88]: I copied that from OTS code, would try to find BaseApi better method

File: cmd/rpcdaemon/commands/graphql_api.go (hunk 3 above)
Comment (SLoeuillet) [reply cmd/rpcdaemon/commands/graphql_api.go:134]: Got that part from OTS code too, not used anymore in my code

**Comment by SLoeuillet** [comment]
Unit tests from geth imported and improved (tests via regexps, also tests chainID query)

**Review by AskAlexSharov — APPROVED** [review]

**Comment by SLoeuillet** [comment]
For Graphq tests to work, rpcdaemon have to be run with --graphql

--- FAIL: TestGraphQLQueryBlock (0.00s)
    graph_test.go:106: could not post: Post "http://localhost:8545/graphql": dial tcp [::1]:8545: connect: connection refused
FAIL
FAIL	github.com/ledgerwatch/erigon/cmd/rpcdaemon/graphql	0.216s
?   	github.com/ledgerwatch/erigon/cmd/rpcdaemon/graphql/graph	[no test files]

**Comment by SLoeuillet** [comment]
Thanks @AskAlexSharov

**Comment by AskAlexSharov** [comment]
@SLoeuillet 
```
--- FAIL: TestGraphQLQueryBlock (0.00s)
graph_test.go:106: could not post: Post "http://localhost:8545/graphql": dial tcp [::1]:8545: connect: connection refused
FAIL
```
Let's turn this test to UnitTest. I don't see why we can't start graphql server inside test. 
For example we have: 
rpc/server_test.go
rpc/http_test.go

**Comment by s1na** [comment]
Hey I see you guys added basic support for GraphQL. I made a PR to change the schema slightly, basically enforcing every number (Int, Long or BigInt) to be formatted as hex. Feel free to chime in: https://github.com/ethereum/execution-apis/pull/389

**Comment by SLoeuillet** [comment]
Thanks @s1na. Would take a look

**Pull request merged**